### ISSUE TITLE
"BAD status: invalid tag name, attributes, or nesting"

### ISSUE THREAD

**Issue by bitwombat** [xournalpp/xournalpp#3388]
First, thanks so much for Xournal++! What a neat project, and so needed in the open source world.

**Affects versions :**
 - OS: Ubuntu 20.04
 - (Linux only) Desktop environment: - X11
 - Which version of libgtk do you use: 3.24.20
 - Version of Xournal++: 1.1.0 / 9d1277de
 - Installation method: .deb from main site 

**Describe the bug**
When I export a big 1257 page PDF, with annotations and annotations from Okular, I get 
"BAD status: invalid tag name, attributes, or nesting" and a 4029 byte result file (should be ~3.3 MB).

**To Reproduce**
1. File->Export as PDF

Unfortunately, it's a copyrighted work, so I cannot ethically send it on. [Here's the resulting file though](https://github.com/xournalpp/xournalpp/files/7147410/haskell-programming-1.0RC4-screen_with_notes_b.pdf)

**Expected behavior**
A complete exported PDF.

**Screenshots of Problem**

**Additional context**
I've tried to reproduce it with two other PDFs, and by paring down my huge PDF. Cannot reproduce. Hopefully the error message is enough.

**Comment by LittleHuba** [comment]
Hi that is unfortunate to hear. Without the original file or a minimal example to reproduce your issue, we can't really help you. This is a wild goose chase without any steps for reproduction.

You can try to split the file into multiple smaller files and export them piece by piece. Merging the PDF files later on is quite easy.

**Comment by bitwombat** [comment]
Whose message is that (xournal's or a library) and what does it mean?

**Comment by bhennion** [comment]
> Whose message is that (xournal's or a library) and what does it mean?

It's a message from `cairo`: [https://cairographics.org/manual/cairo-Error-handling.html](https://cairographics.org/manual/cairo-Error-handling.html)
Unfortunately, the documentation is quite short. Difficult to pinpoint where the issue comes from.

I couldn't reproduce with a 1183 pages long pdf of 4.3MB (just a bunch of `lipsum[1-150]` in `latex`), with some annotations from Okular and some from Xournal++.

If you split your pdf (say with `pdfjam foo.pdf 1-600` and `pdfjam foo.pdf 601-1257`), do **both** the resulting pdf export well from Xournal++?

**Comment by bitwombat** [comment]
Well, if I send the document through pdfjam, the annotations are removed, and no error from xournalpp when exporting PDF.

If I "Print as PDF" from Chromium, the annotations are all messed up (highlights on top of text, etc.), and no error from xournalpp when exporting PDF.

What a mess!

You could clearly consider this "not xournalpp's problem". It's the only tool that's showing an error (the others just do the wrong thing).

In the interest of open source, here's the file for testing purposes. I'll delete it after you acknowledge receipt.
http://eit.bitwombat.com.au/drop/h.pdf

Let me know what else I can do to help.

**Comment by rolandlo** [comment]
@bitwombat Interestingly after removing the first page with `pdftk`
```
pdftk h.pdf cat 2-end h_tk.pdf
```
the export works fine.

**Comment by bitwombat** [comment]
Are the annotations intact (around page 650)?

I'm guessing removing any page does the same thing, because pdftk is re-processing the whole file and removing whatever inconsistency is in it?

This is what makes me distrust fancy tools for notes :)  (okular, not xournalpp)

**Comment by rolandlo** [comment]
> Are the annotations intact (around page 650)?
> 
No, they are not. The highlighted region on page 651 is exported without text.

**Comment by Technius** [comment]
Does the error occur if you try to export the PDF without annotating it with Okular first?

**Comment by bitwombat** [comment]
Yes!
I don't know why I'm excited by that :)

**Comment by LittleHuba** [comment]
Pease run Xournal++ from console with `G_MESSAGES_DEBUG=Poppler xournalpp`. Then reproduce your problem and post the output from the console here.

**Comment by LittleHuba** [comment]
You can use `pdfunite <source> <destination>` to fix such files for now. This command should not modify your PDF.

**Comment by bitwombat** [comment]
> You can use `pdfunite <source> <destination>` to fix such files for now. This command should not modify your PDF.

That excludes all annotations.

**Comment by bitwombat** [comment]
```
** Message: 15:04:38.454: TEXTDOMAINDIR = (null), PACKAGE_LOCALE_DIR = /usr/share/locale, chosen directory = /usr/share/locale
ALSA lib pcm_dsnoop.c:641:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dmix.c:1089:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
ALSA lib pcm_oss.c:377:(_snd_pcm_oss_open) Unknown field port
ALSA lib pcm_oss.c:377:(_snd_pcm_oss_open) Unknown field port
ALSA lib pcm_usb_stream.c:486:(_snd_pcm_usb_stream_open) Invalid type for card
ALSA lib pcm_usb_stream.c:486:(_snd_pcm_usb_stream_open) Invalid type for card
ALSA lib pcm_dmix.c:1089:(snd_pcm_dmix_open) unable to open slave
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
** Message: 15:04:38.674: Plugin "MigrateFontSizes" UI initialized
BAD status: invalid tag name, attributes, or nesting
Poppler-INFO: 15:04:43.933: Internal error at position -1: cairo context error: invalid tag name, attributes, or nesting<0a>
Poppler-INFO: 15:04:43.933: Internal error at position -1: cairo context error: invalid tag name, attributes, or nesting<0a>
Poppler-INFO: 15:04:43.934: Internal error at position -1: cairo context error: invalid tag name, attributes, or nesting<0a>
Poppler-INFO: 15:04:43.934: Internal error at position -1: cairo context error: invalid tag name, attributes, or nesting<0a>
<thousands>
```

**Comment by LittleHuba** [comment]
You will have to open a ticket at the bug tracker of Poppler. There is nothing further we can do. Please link the ticket here.

**Comment by bitwombat** [comment]
Poppler, 633 open tickets, I don't think I'll bother.
They'll likely point at Okular as the culprit, 915 open bugs ;)

**Comment by LittleHuba** [comment]
> They'll likely point at Okular as the culprit

Which is most likely the culprit^^ 

Devs normally give tickets concerning invalid files quite a high priority. So you actually have a good chance of getting prioritized.

**Comment by bitwombat** [comment]
https://bugs.kde.org/show_bug.cgi?id=442975

**Comment by Photon89** [comment]
> Which is most likely the culprit^^

#3394 is marked as duplicate of this, but I experience the bug with a one page PDF file which has not been touched with Okular. So, possibly a Poppler issue nevertheless?

**Comment by LittleHuba** [comment]
As mentioned in https://bugs.kde.org/show_bug.cgi?id=442975 the bug is most likely in Poppler. If you can share this file, please do so over at https://bugs.kde.org/show_bug.cgi?id=442975. Currently they can't do anything because they are missing a file that causes this bug.

**Comment by Photon89** [comment]
I do have an example file, but I've never touched it with Okular, are you sure that the guys at the KDE bug tracker want to see it? :smile:

**Comment by LittleHuba** [comment]
In that case, open a ticket over at https://gitlab.freedesktop.org/poppler/poppler/ if there is not already one open for this issue.

**Comment by Photon89** [comment]
@LittleHuba I see, thanks for the hint! I tried to get some kind of output to present upstream, but I don't get any output from Xournal++ upon pressing the PDF export button. Is there some kind of "--verbose" option which is not mentioned in the "--help" output?

**Comment by LittleHuba** [comment]
Start Xournal++ with `G_MESSAGES_DEBUG=Poppler xournalpp` to receive debug messages of Poppler on the console. That is all we can provide.

**Comment by Photon89** [comment]
Thanks, I think, this should be enough! I get the same four lines as bitwombat. Let's see what upstream says about it!

**Comment by Photon89** [comment]
Upstream report: https://gitlab.freedesktop.org/poppler/poppler/-/issues/1170

**Comment by Photon89** [comment]
Even further upstream report: https://gitlab.freedesktop.org/cairo/cairo/-/issues/523

**Comment by psychon** [comment]
Could one of the affected people help debugging this? Please create a file `/tmp/foo` with the following content:
```
set pagination off
break main
run
delete
break _cairo_error
commands
bt
cont
end
cont
```
Then run `gdb -x xournalpp` (No idea whether `xournalpp` is the binary name of xournal++, you hopefully get the idea).

Sadly, on my Debian testing (and testing with evince), this reports during startup (but then continues anyway):
```
Function "_cairo_error" not defined.
Make breakpoint pending on future shared library load? (y or [n]) [answered N; input not from terminal]
No breakpoint number 1.
```
For my system, installing `libcairo2-dbgsym` made this error go away. Other distros likely have different ways to get debug symbols. I cannot really help you with that.

What the above does: It tells `gdb` to stop execution whenever the function `_cairo_error` is reached and then print a backtrace describing "how we got here", then continue execution. Hopefully, the output of `gdb` can shed some light on which piece of software is at fault here.

-----

Poppler devs said:

>  The `invalid tag name, attributes, or nesting` error message from cairo should only be possible when using the cairo tag API. Poppler does not use this API.

With  my cairo head on: This is not an error in cairo, it is an error that cairo reports that something is mis-using the `cairo_tag()` API. So, not a bug in cairo either.

If you are building cairo from git, you can set the environment variable `CAIRO_DEBUG_TAG=1` to have cairo print more details about tag errors. This is only available since commit 24616585ecb from July and I am not too sure whether the result would really be helpful. The backtrace described above would be a much more useful.

**Comment by Photon89** [comment]
Thanks for the detailed instructions! I will try to rebuild the xournalpp with debugging symbols and see if I can get a gdb backtrace.

**Comment by psychon** [comment]
Uhm... I expect that the problem will be in librsvg. It appears in the build dependencies on https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md and it calls cairo's tag API according to https://codesearch.debian.net/search?q=cairo_tag_begin&literal=1. In fact, it is basically the only option from that list, I think....

**Comment by Photon89** [comment]
Hmm, I just tried downgrading librsvg to several older versions (2.52, 2.50 and 2.48) but the problem persists.

**Comment by Photon89** [comment]
@psychon Something goes very wrong:

```
$ gdb -x /usr/bin/xournalpp
GNU gdb (GDB) 11.1
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
/usr/bin/xournalpp:1: Error in sourced command file:
Undefined command: "".  Try "help".
(gdb)
```

If I run without the `-x` option, I can run Xournal++ from inside gdb but a `bt` gives just `No stack.`.

**Comment by psychon** [comment]
```
Function "_cairo_error" not defined.
Make breakpoint pending on future shared library load? (y or [n]) [answered N; input not from terminal]
No breakpoint number 1.
```
As written above: Missing debugging symbols for cairo :-(

Also: Sorry for forgetting the `/tmp/foo` argument and thanks for figuring that out.

**Comment by Photon89** [comment]
Now it gives me the following:

```
$ LANG=C gdb -x /tmp/foo xournalpp
GNU gdb (GDB) 11.1
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from xournalpp...
Breakpoint 1 at 0xa7430: file /usr/src/debug/xournalpp/src/Xournalpp.cpp, line 23.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".

Breakpoint 1, main (argc=1, argv=0x7fffffffe6e8) at /usr/src/debug/xournalpp/src/Xournalpp.cpp:23
warning: Source file is more recent than executable.
23	auto main(int argc, char* argv[]) -> int {
Breakpoint 2 at 0x7ffff72e64f0: _cairo_error. (24 locations)
[New Thread 0x7ffff200f640 (LWP 43829)]
[New Thread 0x7ffff180e640 (LWP 43830)]
** Message: 17:23:46.357: TEXTDOMAINDIR = (null), PACKAGE_LOCALE_DIR = /usr/share/locale, chosen directory = /usr/share/locale
ALSA lib pcm_dmix.c:1035:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
[New Thread 0x7ffff0f1a640 (LWP 43831)]
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
[Thread 0x7ffff0f1a640 (LWP 43831) exited]
[New Thread 0x7ffff0f1a640 (LWP 43832)]
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
[Thread 0x7ffff0f1a640 (LWP 43832) exited]
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
[New Thread 0x7ffff0a8e640 (LWP 43833)]
[New Thread 0x7fffe3fff640 (LWP 43834)]
[Thread 0x7fffe3fff640 (LWP 43834) exited]
[Thread 0x7ffff0a8e640 (LWP 43833) exited]
[New Thread 0x7ffff0a8e640 (LWP 43835)]
[New Thread 0x7fffe3fff640 (LWP 43836)]
[Thread 0x7fffe3fff640 (LWP 43836) exited]
[Thread 0x7ffff0a8e640 (LWP 43835) exited]
[New Thread 0x7ffff0a8e640 (LWP 43837)]
[Thread 0x7ffff0a8e640 (LWP 43837) exited]
[New Thread 0x7ffff0a8e640 (LWP 43838)]
[Thread 0x7ffff0a8e640 (LWP 43838) exited]
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib pcm_dmix.c:1035:(snd_pcm_dmix_open) unable to open slave
[New Thread 0x7ffff0a8e640 (LWP 43839)]
[Thread 0x7ffff0a8e640 (LWP 43839) exited]
[New Thread 0x7ffff0a8e640 (LWP 43840)]
[Thread 0x7ffff0a8e640 (LWP 43840) exited]
[New Thread 0x7ffff0f1a640 (LWP 43841)]
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
[Thread 0x7ffff0f1a640 (LWP 43841) exited]
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
** Message: 17:23:47.458: Plugin "MigrateFontSizes" UI initialized
[New Thread 0x7ffff0a8e640 (LWP 43842)]
[Thread 0x7ffff0a8e640 (LWP 43842) exited]
[New Thread 0x7ffff0a8e640 (LWP 43843)]
[New Thread 0x7fffe3fff640 (LWP 43844)]
[Thread 0x7ffff0a8e640 (LWP 43843) exited]
[Thread 0x7fffe3fff640 (LWP 43844) exited]
[New Thread 0x7fffe3fff640 (LWP 43845)]
[New Thread 0x7ffff0a8e640 (LWP 43846)]
[Thread 0x7fffe3fff640 (LWP 43845) exited]
[Thread 0x7ffff0a8e640 (LWP 43846) exited]
[New Thread 0x7ffff0a8e640 (LWP 43847)]
[New Thread 0x7fffe3fff640 (LWP 43848)]
[New Thread 0x7fffe3130640 (LWP 43849)]
[Thread 0x7fffe3fff640 (LWP 43848) exited]
[Thread 0x7fffe3130640 (LWP 43849) exited]
[New Thread 0x7fffe3130640 (LWP 43850)]
[New Thread 0x7fffe3fff640 (LWP 43851)]
[Thread 0x7fffe3130640 (LWP 43850) exited]
[New Thread 0x7fffe3fff640 (LWP 43852)]
[Thread 0x7fffe3fff640 (LWP 43851) exited]
[New Thread 0x7fffe3130640 (LWP 43853)]
[Thread 0x7fffe3fff640 (LWP 43852) exited]
[Thread 0x7fffe3130640 (LWP 43853) exited]
[New Thread 0x7fffe3130640 (LWP 43855)]
[New Thread 0x7fffe3fff640 (LWP 43856)]
[Thread 0x7fffe3130640 (LWP 43855) exited]
[New Thread 0x7fffe3130640 (LWP 43857)]
[Thread 0x7fffe3fff640 (LWP 43856) exited]
[New Thread 0x7fffe3fff640 (LWP 43858)]
[Thread 0x7fffe3130640 (LWP 43857) exited]
[Thread 0x7fffe3fff640 (LWP 43858) exited]
[New Thread 0x7fffe3fff640 (LWP 43859)]
[New Thread 0x7fffe3130640 (LWP 43863)]
[New Thread 0x7fffe17da640 (LWP 43864)]
[New Thread 0x7fffe0fd9640 (LWP 43865)]
[New Thread 0x7fffd3fff640 (LWP 43866)]
[New Thread 0x7fffd37fe640 (LWP 43867)]
[New Thread 0x7fffd2ffd640 (LWP 43868)]
[New Thread 0x7fffd27fc640 (LWP 43869)]
[New Thread 0x7fffd1ffb640 (LWP 43870)]
[New Thread 0x7fffd17fa640 (LWP 43871)]
[Thread 0x7fffd1ffb640 (LWP 43870) exited]
[New Thread 0x7fffd17fa640 (LWP 43872)]
[Thread 0x7fffd17fa640 (LWP 43871) exited]
[New Thread 0x7fffd1ffb640 (LWP 43873)]
[Thread 0x7fffd17fa640 (LWP 43872) exited]
[Thread 0x7fffd1ffb640 (LWP 43873) exited]
[Thread 0x7fffe17da640 (LWP 43864) exited]
[Thread 0x7fffd37fe640 (LWP 43867) exited]
[Thread 0x7fffd2ffd640 (LWP 43868) exited]
[Thread 0x7ffff0a8e640 (LWP 43847) exited]
[Thread 0x7fffe0fd9640 (LWP 43865) exited]
[Thread 0x7fffd3fff640 (LWP 43866) exited]
[New Thread 0x7fffd3fff640 (LWP 43880)]
[New Thread 0x7fffe0fd9640 (LWP 43881)]
[New Thread 0x7ffff0a8e640 (LWP 43882)]
[New Thread 0x7fffd2ffd640 (LWP 43883)]
[New Thread 0x7fffe17da640 (LWP 43884)]
[New Thread 0x7fffd37fe640 (LWP 43885)]
[Thread 0x7fffd37fe640 (LWP 43885) exited]
[Thread 0x7fffe0fd9640 (LWP 43881) exited]
[Thread 0x7fffe17da640 (LWP 43884) exited]
[Thread 0x7fffd3fff640 (LWP 43880) exited]
[Thread 0x7fffe3130640 (LWP 43863) exited]
[Thread 0x7ffff0a8e640 (LWP 43882) exited]
** Message: 17:23:57.000: lastOpenPath set
[New Thread 0x7ffff0a8e640 (LWP 43886)]
[New Thread 0x7fffe3130640 (LWP 43887)]
[New Thread 0x7ffff0a8e640 (LWP 43888)]
[Thread 0x7fffe3130640 (LWP 43887) exited]
[Thread 0x7ffff0a8e640 (LWP 43886) exited]
[New Thread 0x7fffe3130640 (LWP 43889)]
[Thread 0x7ffff0a8e640 (LWP 43888) exited]
[Thread 0x7fffe3130640 (LWP 43889) exited]
[New Thread 0x7fffe3130640 (LWP 43890)]
[New Thread 0x7ffff0a8e640 (LWP 43891)]
[New Thread 0x7fffd3fff640 (LWP 43892)]
[New Thread 0x7fffe17da640 (LWP 43893)]
[New Thread 0x7fffe22ea640 (LWP 43894)]
[New Thread 0x7fffe0fd9640 (LWP 43895)]
[Thread 0x7fffe22ea640 (LWP 43894) exited]
[Thread 0x7fffe0fd9640 (LWP 43895) exited]
[Thread 0x7fffe17da640 (LWP 43893) exited]
[Thread 0x7ffff0a8e640 (LWP 43891) exited]
[Thread 0x7fffe3130640 (LWP 43890) exited]
[Thread 0x7fffd2ffd640 (LWP 43883) exited]
[Thread 0x7fffe3fff640 (LWP 43859) exited]
[Thread 0x7fffd3fff640 (LWP 43892) exited]
[Thread 0x7fffd27fc640 (LWP 43869) exited]
[Thread 0x7ffff180e640 (LWP 43830) exited]
[Thread 0x7ffff200f640 (LWP 43829) exited]
[Inferior 1 (process 43825) exited normally]
(gdb) quit
```

**Comment by psychon** [comment]
Hrm... Does this still produce an empty PDF? Could you try adding `G_MESSAGES_DEBUG=Poppler` to the front of this invocation just to make sure this still hits the same problem?

If that doesn't work... I guess I'll have to find some time to try to reproduce this. Assuming this even reproduces on Debian/testing...

**Comment by Photon89** [comment]
> Hrm... Does this still produce an empty PDF? 

Ha! No, it doesn't any more. If I run with German locale, it produces an empty PDF; if I run with LANG=C, it doesn't! So let me run gdb with German locale:
```

$ gdb -x /tmp/foo xournalpp
GNU gdb (GDB) 11.1
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from xournalpp...
Breakpoint 1 at 0xa7430: file /usr/src/debug/xournalpp/src/Xournalpp.cpp, line 23.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".

Breakpoint 1, main (argc=1, argv=0x7fffffffe6d8) at /usr/src/debug/xournalpp/src/Xournalpp.cpp:23
warning: Source file is more recent than executable.
23	auto main(int argc, char* argv[]) -> int {
Breakpoint 2 at 0x7ffff72e64f0: _cairo_error. (24 locations)
[New Thread 0x7ffff1c36640 (LWP 44878)]
[New Thread 0x7ffff1435640 (LWP 44879)]
** Message: 17:49:40.184: TEXTDOMAINDIR = (null), PACKAGE_LOCALE_DIR = /usr/share/locale, chosen directory = /usr/share/locale
ALSA lib pcm_dmix.c:1035:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
[New Thread 0x7ffff0aef640 (LWP 44880)]
Cannot connect to server socket err = Datei oder Verzeichnis nicht gefunden
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
[Thread 0x7ffff0aef640 (LWP 44880) exited]
[New Thread 0x7ffff0aef640 (LWP 44881)]
Cannot connect to server socket err = Datei oder Verzeichnis nicht gefunden
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
[Thread 0x7ffff0aef640 (LWP 44881) exited]
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
[New Thread 0x7fffe3fff640 (LWP 44882)]
[New Thread 0x7fffe37fe640 (LWP 44883)]
[Thread 0x7fffe37fe640 (LWP 44883) exited]
[Thread 0x7fffe3fff640 (LWP 44882) exited]
[New Thread 0x7fffe3fff640 (LWP 44884)]
[New Thread 0x7fffe37fe640 (LWP 44885)]
[Thread 0x7fffe37fe640 (LWP 44885) exited]
[Thread 0x7fffe3fff640 (LWP 44884) exited]
[New Thread 0x7fffe3fff640 (LWP 44886)]
[Thread 0x7fffe3fff640 (LWP 44886) exited]
[New Thread 0x7fffe3fff640 (LWP 44887)]
[Thread 0x7fffe3fff640 (LWP 44887) exited]
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib pcm_dmix.c:1035:(snd_pcm_dmix_open) unable to open slave
[New Thread 0x7fffe3fff640 (LWP 44888)]
[Thread 0x7fffe3fff640 (LWP 44888) exited]
[New Thread 0x7fffe3fff640 (LWP 44889)]
[Thread 0x7fffe3fff640 (LWP 44889) exited]
[New Thread 0x7ffff0aef640 (LWP 44890)]
Cannot connect to server socket err = Datei oder Verzeichnis nicht gefunden
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
[Thread 0x7ffff0aef640 (LWP 44890) exited]
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
** Message: 17:49:41.268: Plugin "MigrateFontSizes" UI initialized
[New Thread 0x7fffe3fff640 (LWP 44891)]
[Thread 0x7fffe3fff640 (LWP 44891) exited]
[New Thread 0x7fffe3fff640 (LWP 44892)]
[New Thread 0x7fffe37fe640 (LWP 44893)]
[Thread 0x7fffe3fff640 (LWP 44892) exited]
[New Thread 0x7fffe3fff640 (LWP 44894)]
[New Thread 0x7fffe2e0a640 (LWP 44895)]
[Thread 0x7fffe2e0a640 (LWP 44895) exited]
[Thread 0x7fffe3fff640 (LWP 44894) exited]
[Thread 0x7fffe37fe640 (LWP 44893) exited]
[New Thread 0x7fffe2e0a640 (LWP 44896)]
[New Thread 0x7fffe3fff640 (LWP 44898)]
[New Thread 0x7fffe37fe640 (LWP 44899)]
[Thread 0x7fffe3fff640 (LWP 44898) exited]
[Thread 0x7fffe37fe640 (LWP 44899) exited]
[New Thread 0x7fffe37fe640 (LWP 44900)]
[New Thread 0x7fffe3fff640 (LWP 44901)]
[Thread 0x7fffe37fe640 (LWP 44900) exited]
[New Thread 0x7fffe37fe640 (LWP 44902)]
[New Thread 0x7fffe2285640 (LWP 44903)]
[Thread 0x7fffe3fff640 (LWP 44901) exited]
[Thread 0x7fffe37fe640 (LWP 44902) exited]
[Thread 0x7fffe2285640 (LWP 44903) exited]
[New Thread 0x7fffe2285640 (LWP 44904)]
[New Thread 0x7fffe37fe640 (LWP 44905)]
[Thread 0x7fffe2285640 (LWP 44904) exited]
[New Thread 0x7fffe2285640 (LWP 44906)]
[Thread 0x7fffe37fe640 (LWP 44905) exited]
[New Thread 0x7fffe37fe640 (LWP 44907)]
[Thread 0x7fffe2285640 (LWP 44906) exited]
[Thread 0x7fffe37fe640 (LWP 44907) exited]
[New Thread 0x7fffe37fe640 (LWP 44908)]
[New Thread 0x7fffe2285640 (LWP 44911)]
[New Thread 0x7fffe3fff640 (LWP 44912)]
[New Thread 0x7fffe0be1640 (LWP 44913)]
[New Thread 0x7fffcffff640 (LWP 44914)]
[New Thread 0x7fffcf7fe640 (LWP 44915)]
[New Thread 0x7fffceffd640 (LWP 44916)]
[New Thread 0x7fffce7fc640 (LWP 44917)]
[New Thread 0x7fffcdffb640 (LWP 44918)]
[New Thread 0x7fffcd7fa640 (LWP 44919)]
[Thread 0x7fffcdffb640 (LWP 44918) exited]
[New Thread 0x7fffcdffb640 (LWP 44920)]
[Thread 0x7fffcd7fa640 (LWP 44919) exited]
[New Thread 0x7fffcd7fa640 (LWP 44921)]
[Thread 0x7fffcdffb640 (LWP 44920) exited]
[Thread 0x7fffcd7fa640 (LWP 44921) exited]
[Thread 0x7fffcffff640 (LWP 44914) exited]
[Thread 0x7fffceffd640 (LWP 44916) exited]
[Thread 0x7fffe3fff640 (LWP 44912) exited]
[Thread 0x7fffe2285640 (LWP 44911) exited]
[Thread 0x7fffe2e0a640 (LWP 44896) exited]
[Thread 0x7fffe0be1640 (LWP 44913) exited]
** Message: 17:49:47.231: lastOpenPath set
[New Thread 0x7fffe0be1640 (LWP 44924)]
[New Thread 0x7fffe2e0a640 (LWP 44925)]
[Thread 0x7fffe0be1640 (LWP 44924) exited]
[New Thread 0x7fffe0be1640 (LWP 44926)]
[New Thread 0x7fffceffd640 (LWP 44927)]
[Thread 0x7fffe2e0a640 (LWP 44925) exited]
[Thread 0x7fffe0be1640 (LWP 44926) exited]
[Thread 0x7fffceffd640 (LWP 44927) exited]
[New Thread 0x7fffceffd640 (LWP 44928)]
[New Thread 0x7fffe0be1640 (LWP 44929)]
[New Thread 0x7fffe2e0a640 (LWP 44930)]
[New Thread 0x7fffe2285640 (LWP 44931)]
[New Thread 0x7fffe16f1640 (LWP 44932)]
[New Thread 0x7fffcffff640 (LWP 44933)]
[Thread 0x7fffe2285640 (LWP 44931) exited]
[Thread 0x7fffe0be1640 (LWP 44929) exited]
[Thread 0x7fffe16f1640 (LWP 44932) exited]
[Thread 0x7fffcffff640 (LWP 44933) exited]
[Thread 0x7fffcf7fe640 (LWP 44915) exited]
[Thread 0x7fffe2e0a640 (LWP 44930) exited]
[Thread 0x7fffe37fe640 (LWP 44908) exited]
[Thread 0x7fffceffd640 (LWP 44928) exited]
[Thread 0x7ffff1435640 (LWP 44879) exited]
[Thread 0x7ffff1c36640 (LWP 44878) exited]
[Thread 0x7ffff2069080 (LWP 44874) exited]
[Inferior 1 (process 44874) exited normally]
```

**Comment by psychon** [comment]
Gut, dass ich auch ein deutsches Locale nutze. ;-)

Hm, funnily enough, there was no warning output from Pango and the breakpoint wasn't hit. I guess I'll have to try to reproduce this when I have some time, but I'm not sure when I'll have time for that

**Comment by psychon** [comment]
> Aber bei deinem System tritt das Problem nicht auf, oder?

xournalpp's build dependencies scared me away from trying this so far. I had hoped `gdb` would provide an easy answer quickly.

-----

I had one more idea: Could you replace `break _cairo_error` with `break cairo_tag_begin`? Instead of trying to figure out "what's the state when the error occurs?", this breakpoint would answer the question "who is calling cairo's tag API?". Since not much code uses this API, this could already be useful.

**Comment by Photon89** [comment]
> xournalpp's build dependencies scared me away from trying this so far. I had hoped gdb would provide an easy answer quickly.

Maybe it is possible to ask the Xournal++ devs for a -dbg package? They have a "normal" package for Debian already: https://github.com/xournalpp/xournalpp/releases/tag/1.1.0

> I had one more idea: Could you replace break _cairo_error with break cairo_tag_begin?

Sure, here you go:

```
$ gdb -x /home/photon/poppler/foo xournalpp
GNU gdb (GDB) 11.1
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from xournalpp...
Breakpoint 1 at 0xa7430: file /usr/src/debug/xournalpp/src/Xournalpp.cpp, line 23.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".

Breakpoint 1, main (argc=1, argv=0x7fffffffe708) at /usr/src/debug/xournalpp/src/Xournalpp.cpp:23
warning: Source file is more recent than executable.
23	auto main(int argc, char* argv[]) -> int {
Breakpoint 2 at 0x7ffff733aae0: file ../cairo/src/cairo.c, line 2938.
[New Thread 0x7ffff1c36640 (LWP 1708)]
[New Thread 0x7ffff1435640 (LWP 1709)]
** Message: 15:08:22.720: TEXTDOMAINDIR = (null), PACKAGE_LOCALE_DIR = /usr/share/locale, chosen directory = /usr/share/locale
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
[New Thread 0x7ffff0aef640 (LWP 1710)]
Cannot connect to server socket err = Datei oder Verzeichnis nicht gefunden
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
[Thread 0x7ffff0aef640 (LWP 1710) exited]
[New Thread 0x7ffff0aef640 (LWP 1711)]
Cannot connect to server socket err = Datei oder Verzeichnis nicht gefunden
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
[Thread 0x7ffff0aef640 (LWP 1711) exited]
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
[New Thread 0x7fffe3fff640 (LWP 1712)]
[New Thread 0x7fffe37fe640 (LWP 1713)]
[Thread 0x7fffe37fe640 (LWP 1713) exited]
[Thread 0x7fffe3fff640 (LWP 1712) exited]
[New Thread 0x7fffe3fff640 (LWP 1716)]
[New Thread 0x7fffe37fe640 (LWP 1717)]
[Thread 0x7fffe37fe640 (LWP 1717) exited]
[Thread 0x7fffe3fff640 (LWP 1716) exited]
[New Thread 0x7fffe3fff640 (LWP 1718)]
[Thread 0x7fffe3fff640 (LWP 1718) exited]
[New Thread 0x7fffe3fff640 (LWP 1719)]
[Thread 0x7fffe3fff640 (LWP 1719) exited]
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
[New Thread 0x7fffe3fff640 (LWP 1720)]
[Thread 0x7fffe3fff640 (LWP 1720) exited]
[New Thread 0x7fffe3fff640 (LWP 1721)]
[Thread 0x7fffe3fff640 (LWP 1721) exited]
[New Thread 0x7ffff0aef640 (LWP 1722)]
Cannot connect to server socket err = Datei oder Verzeichnis nicht gefunden
Cannot connect to server request channel
jack server is not running or cannot be started
[Thread 0x7ffff0aef640 (LWP 1722) exited]
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
** Message: 15:08:23.867: Plugin "MigrateFontSizes" UI initialized
[New Thread 0x7fffe3fff640 (LWP 1723)]
[Thread 0x7fffe3fff640 (LWP 1723) exited]
[New Thread 0x7fffe3fff640 (LWP 1724)]
[New Thread 0x7fffe37fe640 (LWP 1725)]
[Thread 0x7fffe3fff640 (LWP 1724) exited]
[New Thread 0x7fffe3fff640 (LWP 1726)]
[Thread 0x7fffe37fe640 (LWP 1725) exited]
[New Thread 0x7fffe37fe640 (LWP 1727)]
[Thread 0x7fffe3fff640 (LWP 1726) exited]
[Thread 0x7fffe37fe640 (LWP 1727) exited]
[New Thread 0x7fffe37fe640 (LWP 1728)]
[New Thread 0x7fffe3fff640 (LWP 1729)]
[New Thread 0x7fffe2742640 (LWP 1730)]
[Thread 0x7fffe3fff640 (LWP 1729) exited]
[Thread 0x7fffe2742640 (LWP 1730) exited]
[New Thread 0x7fffe2742640 (LWP 1731)]
[New Thread 0x7fffe3fff640 (LWP 1732)]
[Thread 0x7fffe2742640 (LWP 1731) exited]
[New Thread 0x7fffe2742640 (LWP 1733)]
[Thread 0x7fffe3fff640 (LWP 1732) exited]
[New Thread 0x7fffe3fff640 (LWP 1734)]
[Thread 0x7fffe2742640 (LWP 1733) exited]
[Thread 0x7fffe3fff640 (LWP 1734) exited]
[New Thread 0x7fffe3fff640 (LWP 1735)]
[New Thread 0x7fffe2742640 (LWP 1736)]
[Thread 0x7fffe3fff640 (LWP 1735) exited]
[New Thread 0x7fffe3fff640 (LWP 1737)]
[Thread 0x7fffe2742640 (LWP 1736) exited]
[New Thread 0x7fffe2742640 (LWP 1738)]
[Thread 0x7fffe3fff640 (LWP 1737) exited]
[Thread 0x7fffe2742640 (LWP 1738) exited]
[New Thread 0x7fffe2742640 (LWP 1739)]
[New Thread 0x7fffe3fff640 (LWP 1741)]
[New Thread 0x7fffe0eac640 (LWP 1742)]
[New Thread 0x7fffcffff640 (LWP 1743)]
[New Thread 0x7fffcf7fe640 (LWP 1744)]
[New Thread 0x7fffceffd640 (LWP 1745)]
[New Thread 0x7fffce7fc640 (LWP 1746)]
[New Thread 0x7fffcdffb640 (LWP 1747)]
[New Thread 0x7fffcd7fa640 (LWP 1748)]
[New Thread 0x7fffccff9640 (LWP 1749)]
[Thread 0x7fffcd7fa640 (LWP 1748) exited]
[New Thread 0x7fffcd7fa640 (LWP 1750)]
[New Thread 0x7fffaffff640 (LWP 1751)]
[Thread 0x7fffccff9640 (LWP 1749) exited]
[Thread 0x7fffcd7fa640 (LWP 1750) exited]
[Thread 0x7fffaffff640 (LWP 1751) exited]
[Thread 0x7fffcffff640 (LWP 1743) exited]
[Thread 0x7fffceffd640 (LWP 1745) exited]
[Thread 0x7fffe0eac640 (LWP 1742) exited]
[Thread 0x7fffe37fe640 (LWP 1728) exited]
[Thread 0x7fffe3fff640 (LWP 1741) exited]
[Thread 0x7fffce7fc640 (LWP 1746) exited]
** Message: 15:08:30.276: lastOpenPath set
[New Thread 0x7fffce7fc640 (LWP 1775)]
[New Thread 0x7fffe3fff640 (LWP 1776)]
[Thread 0x7fffce7fc640 (LWP 1775) exited]
[Thread 0x7fffe3fff640 (LWP 1776) exited]
[New Thread 0x7fffe3fff640 (LWP 1777)]
[New Thread 0x7fffce7fc640 (LWP 1778)]
[Thread 0x7fffe3fff640 (LWP 1777) exited]
[Thread 0x7fffce7fc640 (LWP 1778) exited]
[New Thread 0x7fffce7fc640 (LWP 1779)]
[New Thread 0x7fffe3fff640 (LWP 1780)]
[New Thread 0x7fffe37fe640 (LWP 1781)]
[New Thread 0x7fffceffd640 (LWP 1782)]
[New Thread 0x7fffe19bc640 (LWP 1783)]
[New Thread 0x7fffe11bb640 (LWP 1784)]
[Thread 0x7fffe19bc640 (LWP 1783) exited]
[Thread 0x7fffce7fc640 (LWP 1779) exited]
[Thread 0x7fffceffd640 (LWP 1782) exited]
[Thread 0x7fffe3fff640 (LWP 1780) exited]
[Thread 0x7fffe37fe640 (LWP 1781) exited]
[Thread 0x7fffcf7fe640 (LWP 1744) exited]
[Thread 0x7fffe2742640 (LWP 1739) exited]
[Thread 0x7fffe11bb640 (LWP 1784) exited]
[Thread 0x7fffcdffb640 (LWP 1747) exited]
[Thread 0x7ffff1435640 (LWP 1709) exited]
[Thread 0x7ffff1c36640 (LWP 1708) exited]
[Inferior 1 (process 1704) exited normally]
(gdb) 
```

**Comment by psychon** [comment]
Urgh.... in other news, nothing is using this cairo API....

Hm... Does something in here start a new process and the error only occurs in the child? If so, `set follow-fork-mode child` would make `gdb` follow the child process instead of the parent, but.... I guess I am reaching for straws here.

Sorry :-(

**Comment by LittleHuba** [comment]
I'll reopen this ticket officially as it is obviously not fixed...

Should you require debug data from within our code and don't want to set up all dependencies, you can always open a PR that creates the necessary debug messages. It is really simple for us to provide you installers for the current state of a PR.

**Comment by Photon89** [comment]
@psychon If you tell me, what I should do, I can test it. :smiley: 

Actually, what I noticed is that the debugging output comes very early when I haven't had a chance to do the PDF export. When I do the export, no output is being added. Maybe the debugging stuff should be paused somehow to wait till the export actually happens?

**Comment by psychon** [comment]
@Photon89 What Linux Distro are you on?

I just build Xournal++ from source, started it, used `lsof -p $(pidof xournalpp) | grep '.so'` to figure out which libraries it uses. For each of them, I used `grep -q cairo` to check if it depends on cairo: `lsof -p 8636 | grep '.so' | cut -c 75-   | while read file ; do if grep -q cairo $file ; then echo $file ; fi ; done`. The resulting list has nine entries. For each of them I checked their git repo and search for `cairo_tag_begin`. No hits. But librsvg2 has a hit for `tag_begin`, but that code looks safe/fine. :-(

Oh and no: Of course I cannot reproduce this problem with my debian testing system.

So... I guess I am out of ideas. No one calls `cairo_tag_begin` and yet one of these zero calls fails...

**Comment by Photon89** [comment]
I'm, on Arch and Manjaro (problem occurs on both machines). I can provide a list or tree (you choose) of dependencies for Xournal++ in Arch/Manjaro, if this helps?

What do you think about the timing issue I mentioned in my previous comment?

**Comment by bhennion** [comment]
I can reproduce on Arch Linux with @Photon89's original files from the [poppler bug report](https://gitlab.freedesktop.org/poppler/poppler/-/issues/1170). I had to edit the .xopp file to get the right background pdf file.

Interestingly, running `xournalpp --create-pdf foo.pdf bar.xopp` works while exporting from within Xournal++ does not. @Photon89, do you get the same behaviour with the CLI?

**Comment by Photon89** [comment]
Yes, I can confirm this behaviour!

**Comment by psychon** [comment]
@bhennion Are you also using a German locale? (`LC_ALL=de_DE.UTF-8`, I guess)

Somehow this locale sounds like something somewhere mixes up `,` and `.` in a floating point number... but... how?

**Comment by bhennion** [comment]
No, the french locale `fr_FR.UTF-8`. But in french like in german, 1.2 is written 1,2, so it does not change anything with respect to your comma theory.

**Comment by Photon89** [comment]
I think, input from Xournal++ devs would be helpful, they know best what differences there are when using the CLI compared to using the GUI.

**Comment by bhennion** [comment]
Well, I am a dev, but I can't find anything that could explain this behaviour.

When called from CLI, the conversion to PDF happens in `src/control/XournalMain.cpp:exportPdf()`
When called from the `Export to PDF` menu item, the conversion happens in `src/control/jobs/PdfExportJob.cpp:run()`.
When called from the `Custom Export` menu item, the conversion happens in `src/control/jobs/CustomExportJob.cpp:run()`

In each one of those cases, it instantiates a `XojPdfExport` object and exports through it.
There aren't a lot of differences between those code sections.

My guess is, the main difference between CLI and GUI is, well, that the GUI is running and the export happens in a thread.

**Comment by Photon89** [comment]
Is it possible to add some debugging lines to better understand what is happening in each of the cases? At least pinpoint the breakage to Cairo, Poppler or Xournal++ itself...

**Comment by bhennion** [comment]
It is surely possible. I don't have a lot of time those days, so it'd go faster if you did it. I'd start by adding debugging outputs everywhere in `XojCairoPdfExport::createPdf` in `src/pdf/base/XojCairoPdfExport.cpp`. This is where everything happens (CLI or GUI alike).

**Comment by Photon89** [comment]
I fear, I'm a bit overwhelmed by such a task, the last time when I worked with C++ was almost ten years ago and I was a complete beginner back then...

**Comment by bhennion** [comment]
No worries! Then I'll have a look when I have time.

**Comment by Photon89** [comment]
Great, thanks! I will use the "Print" dialog to get a PDF "printout" as workaround in the mean time.

**Comment by bhennion** [comment]
Ok, I found some time :-)

It turns out the issue comes from... Portaudio! Adding a simple line
```
portaudio::AutoSystem autoSys;
```
to the beginning of `src/control/XournalMain.cpp:exportPdf()` makes the CLI exports to fail with the same error messages as with the GUI exports.

I assume Portaudio sets up an environment variable that poppler/cairo/librsvg uses, and something fails.

**Comment by bhennion** [comment]
Also, there are no error if we comment out exporting the table of content (though calls to `cairo_pdf_surface_add_outline`).
I don't know how this works yet, so it's hard for me to build a minimal working example.

**Comment by bhennion** [comment]
Well, this is a mindf***, but a fix is available #3551.

The fix consists in making sure the locale used for writing PDF table of content metadata is the classic locale (i.e. `C`).

I don't know why `portaudio` changes anything in the story (it does not change the locale), but I don't think it matters.

**Comment by psychon** [comment]
Argh. Sorry for totally being on the wrong track. I really did not know that `cairo_pdf_surface_add_outline` can cause cairo tag errors. At least it makes sense that the locale is involved in this and the floating point numbers really are affected.

I would suggest to also add a check like `if (cairo_surface_status(this->surface) != CAIRO_STATUS_SUCCESS) { print some message that makes it easier to spot where the error comes from if this ever happens again }`

> I don't know why portaudio changes anything in the story (it does not change the locale), but I don't think it matters.

A random guess would be: "Calls `std::locale::global`", but I couldn't find any such call.

Oh, wait. I just did:

https://github.com/xournalpp/xournalpp/blob/e6bb8b797fa5ca66c55724bed5017863b7a0f36c/src/control/XournalMain.cpp#L89

No idea why portaudio is involved, but at least now I understand why all streams end up with the locale. But yeah, I'm also done.

**Comment by bhennion** [comment]
> Argh. Sorry for totally being on the wrong track. I really did not know that `cairo_pdf_surface_add_outline` can cause cairo tag errors. At least it makes sense that the locale is involved in this and the floating point numbers really are affected.

Not sure `cairo_pdf_surface_add_outline` directly caused the tag errors. It seems like we entered the realm of undefined behaviour with those commas in link attributes.
@Photon89's file had link attributes of the form `page=2 pos=[70 74,92]` and this triggered the tag errors later, when `poppler_page_render_for_printing` was called.
Another document with a TOC had link attributes of the form `page=1 pos=[124,802 385,831]` and this did not prompt any error from Poppler (but the resulting pdf was also empty). 

Anyway, thanks a lot for your help!

**Comment by Photon89** [comment]
Great, thanks for the quick fix! If this needs some testing before being accepted, I'd try to build it. Let me know!

**Comment by bhennion** [comment]
> Great, thanks for the quick fix! If this needs some testing before being accepted, I'd try to build it. Let me know!

NP. Yes, it'd be great of you could build the PR and check it works also on your setup.

**Comment by bhennion** [comment]
I found the core source of the issue: the line
```
portaudio::AutoSystem autoSys;
```
 changes the C-style locale (the one you get from `setlocale(LC_NUMERIC, NULL)`). Since `cairo` parses the tag using `sscanf` (see [this line](https://github.com/marionebl/cairo/blob/bc5d3ebe4145b5c1c14d498510dd951f0b1f78b7/src/cairo-tag-attributes.c#L221)), the locale of our `stringstream` and the one `cairo` uses differ.

I see two issues here:

1. `portaudio` should not be changing the locale
2. why are we passing a string to `cairo` for it to be parsed right after? It sounds like this could be avoided...

**Comment by bhennion** [comment]
See upstream [bug report](https://github.com/PortAudio/portaudio/issues/661)

**Comment by bhennion** [comment]
Fixed via #3551 via #3599
